File: builtins/skills/readwrite-discord/scripts/discord.ts
#!/usr/bin/env bun
import { Client, GatewayIntentBits, ChannelType } from "discord.js";

const DISCORD_BOT_TOKEN = process.env.DISCORD_BOT_TOKEN;

if (!DISCORD_BOT_TOKEN) {
  console.log(JSON.stringify({ success: false, error: "DISCORD_BOT_TOKEN not set" }));
  process.exit(1);
}

function parseArgs(args: string[]): Record<string, string> {
  const parsed: Record<string, string> = {};
  for (const arg of args) {
    const match = arg.match(/^--([^=]+)=(.*)$/);
    if (match && match[1] && match[2]) {
      parsed[match[1]] = match[2];
    }
  }
  return parsed;
}

function output(data: unknown): void {
  console.log(JSON.stringify({ success: true, data }));
}

function error(message: string): void {
  console.log(JSON.stringify({ success: false, error: message }));
  process.exit(1);
}

async function withClient<T>(fn: (client: Client) => Promise<T>): Promise<T> {
  const client = new Client({
    intents: [
      GatewayIntentBits.Guilds,
      GatewayIntentBits.GuildMessages,
      GatewayIntentBits.GuildMembers,
      GatewayIntentBits.MessageContent,
    ],
  });

  try {
    await client.login(DISCORD_BOT_TOKEN);
    await new Promise<void>((resolve) => {
      client.once("ready", () => resolve());
    });
    return await fn(client);
  } finally {
    client.destroy();
  }
}

const command = process.argv[2];
const opts = parseArgs(process.argv.slice(3));

async function main(): Promise<void> {
  switch (command) {
    case "list-channels": {
      const guildId = opts["guild-id"];
      if (!guildId) error("--guild-id is required");
      await withClient(async (client) => {
        const guild = await client.guilds.fetch(guildId!);
        const channels = await guild.channels.fetch();
        output(
          channels
            .filter((c) => c !== null)
            .map((c) => ({
              id: c!.id,
              name: c!.name,
              type: ChannelType[c!.type],
              parentId: c!.parentId,
            }))
        );
      });
      break;
    }

    case "read-messages": {
      const channelId = opts["channel-id"];
      if (!channelId) error("--channel-id is required");
      const limit = Math.min(Number(opts["limit"] ?? "25"), 100);
      await withClient(async (client) => {
        const channel = await client.channels.fetch(channelId!);
        if (!channel || !channel.isTextBased()) error("Channel not found or not text-based");
        const messages = await (channel as any).messages.fetch({
          limit,
          ...(opts["before"] ? { before: opts["before"] } : {}),
        });
        output(
          messages.map((m: any) => ({
            id: m.id,
            author: { id: m.author.id, username: m.author.username },
            content: m.content,
            timestamp: m.createdTimestamp,
          }))
        );
      });
      break;
    }

    case "list-members": {
      const guildId = opts["guild-id"];
      if (!guildId) error("--guild-id is required");
      const limit = Math.min(Number(opts["limit"] ?? "100"), 1000);
      await withClient(async (client) => {
        const guild = await client.guilds.fetch(guildId!);
        const members = await guild.members.fetch({ limit });
        output(
          members.map((m) => ({
            id: m.id,
            username: m.user.username,
            displayName: m.displayName,
            roles: m.roles.cache.map((r) => r.name),
          }))
        );
      });
      break;
    }

    case "get-guild": {
      const guildId = opts["guild-id"];
      if (!guildId) error("--guild-id is required");
      await withClient(async (client) => {
        const guild = await client.guilds.fetch(guildId!);
        output({
          id: guild.id,
          name: guild.name,
          memberCount: guild.memberCount,
          ownerId: guild.ownerId,
          createdAt: guild.createdTimestamp,
        });
      });
      break;
    }

    case "get-channel": {
      const channelId = opts["channel-id"];
      if (!channelId) error("--channel-id is required");
      await withClient(async (client) => {
        const channel = await client.channels.fetch(channelId!);
        if (!channel) error("Channel not found");
        output({
          id: channel!.id,
          type: ChannelType[channel!.type],
          ...(("name" in channel!) ? { name: (channel as any).name } : {}),
          ...(("topic" in channel!) ? { topic: (channel as any).topic } : {}),
          ...(("parentId" in channel!) ? { parentId: (channel as any).parentId } : {}),
        });
      });
      break;
    }

    case "send-message": {
      const channelId = opts["channel-id"];
      const content = opts["content"];
      if (!channelId) error("--channel-id is required");
      if (!content) error("--content is required");
      await withClient(async (client) => {
        const channel = await client.channels.fetch(channelId!);
        if (!channel || !channel.isTextBased()) error("Channel not found or not text-based");
        const msg = await (channel as any).send(content);
        output({ id: msg.id, channelId: msg.channelId });
      });
      break;
    }

    case "create-channel": {
      const guildId = opts["guild-id"];
      const name = opts["name"];
      if (!guildId) error("--guild-id is required");
      if (!name) error("--name is required");
      await withClient(async (client) => {
        const guild = await client.guilds.fetch(guildId!);
        const channel = await guild.channels.create({
          name: name!,
          type: ChannelType.GuildText,
          ...(opts["topic"] ? { topic: opts["topic"] } : {}),
          ...(opts["category-id"] ? { parent: opts["category-id"] } : {}),
        });
        output({ id: channel.id, name: channel.name });
      });
      break;
    }

    case "modify-channel": {
      const channelId = opts["channel-id"];
      if (!channelId) error("--channel-id is required");
      await withClient(async (client) => {
        const channel = await client.channels.fetch(channelId!);
        if (!channel) error("Channel not found");
        const updated = await (channel as any).edit({
          ...(opts["name"] ? { name: opts["name"] } : {}),
          ...(opts["topic"] ? { topic: opts["topic"] } : {}),
        });
        output({ id: updated.id, name: updated.name });
      });
      break;
    }

    case "delete-channel": {
      const channelId = opts["channel-id"];
      if (!channelId) error("--channel-id is required");
      await withClient(async (client) => {
        const channel = await client.channels.fetch(channelId!);
        if (!channel) error("Channel not found");
        await (channel as any).delete();
        output({ deleted: channelId });
      });
      break;
    }

    case "add-reaction": {
      const channelId = opts["channel-id"];
      const messageId = opts["message-id"];
      const emoji = opts["emoji"];
      if (!channelId) error("--channel-id is required");
      if (!messageId) error("--message-id is required");
      if (!emoji) error("--emoji is required");
      await withClient(async (client) => {
        const channel = await client.channels.fetch(channelId!);
        if (!channel || !channel.isTextBased()) error("Channel not found or not text-based");
        const message = await (channel as any).messages.fetch(messageId);
        await message.react(emoji);
        output({ reacted: true, messageId, emoji });
      });
      break;
    }

    default:
      error(`Unknown command: ${command}. Available: list-channels, read-messages, list-members, get-guild, get-channel, send-message, create-channel, modify-channel, delete-channel, add-reaction`);
  }
}

main().catch((err) => {
  error(err instanceof Error ? err.message : String(err));
});

File: builtins/skills/read-discord/scripts/discord.py
#!/usr/bin/env python3
"""Discord bot CLI for reading server data via discord.py."""

import asyncio
import json
import os
import sys

import discord


DISCORD_BOT_TOKEN = os.environ.get("DISCORD_BOT_TOKEN")

if not DISCORD_BOT_TOKEN:
    print(json.dumps({"success": False, "error": "DISCORD_BOT_TOKEN not set"}))
    sys.exit(1)


def parse_args(args: list[str]) -> dict[str, str]:
    parsed = {}
    for arg in args:
        if arg.startswith("--") and "=" in arg:
            key, value = arg[2:].split("=", 1)
            parsed[key] = value
    return parsed


def output(data: object) -> None:
    print(json.dumps({"success": True, "data": data}))


def error(message: str) -> None:
    print(json.dumps({"success": False, "error": message}))
    sys.exit(1)


async def with_client(fn):
    intents = discord.Intents.default()
    intents.members = True
    intents.message_content = True
    client = discord.Client(intents=intents)

    ready = asyncio.Event()

    @client.event
    async def on_ready():
        ready.set()

    await client.login(DISCORD_BOT_TOKEN)
    asyncio.ensure_future(client.connect())
    await ready.wait()
    try:
        return await fn(client)
    finally:
        await client.close()


async def main() -> None:
    if len(sys.argv) < 2:
        error("No command specified")

    command = sys.argv[1]
    opts = parse_args(sys.argv[2:])

    match command:
        case "list-channels":
            guild_id = opts.get("guild-id")
            if not guild_id:
                error("--guild-id is required")

            async def run(client: discord.Client):
                guild = await client.fetch_guild(int(guild_id))
                channels = await guild.fetch_channels()
                output([
                    {
                        "id": str(c.id),
                        "name": c.name,
                        "type": str(c.type),
                        "parentId": str(c.category_id) if c.category_id else None,
                    }
                    for c in channels
                ])

            await with_client(run)

        case "read-messages":
            channel_id = opts.get("channel-id")
            if not channel_id:
                error("--channel-id is required")
            limit = min(int(opts.get("limit", "25")), 100)
            before_id = opts.get("before")

            async def run(client: discord.Client):
                channel = await client.fetch_channel(int(channel_id))
                if not isinstance(channel, discord.abc.Messageable):
                    error("Channel not found or not text-based")
                kwargs = {"limit": limit}
                if before_id:
                    kwargs["before"] = discord.Object(id=int(before_id))
                messages = [m async for m in channel.history(**kwargs)]
                output([
                    {
                        "id": str(m.id),
                        "author": {"id": str(m.author.id), "username": m.author.name},
                        "content": m.content,
                        "timestamp": int(m.created_at.timestamp() * 1000),
                    }
                    for m in messages
                ])

            await with_client(run)

        case "list-members":
            guild_id = opts.get("guild-id")
            if not guild_id:
                error("--guild-id is required")
            limit = min(int(opts.get("limit", "100")), 1000)

            async def run(client: discord.Client):
                guild = await client.fetch_guild(int(guild_id))
                members = [m async for m in guild.fetch_members(limit=limit)]
                output([
                    {
                        "id": str(m.id),
                        "username": m.name,
                        "displayName": m.display_name,
                        "roles": [r.name for r in m.roles],
                    }
                    for m in members
                ])

            await with_client(run)

        case "get-guild":
            guild_id = opts.get("guild-id")
            if not guild_id:
                error("--guild-id is required")

            async def run(client: discord.Client):
                guild = await client.fetch_guild(int(guild_id))
                output({
                    "id": str(guild.id),
                    "name": guild.name,
                    "memberCount": guild.approximate_member_count,
                    "ownerId": str(guild.owner_id),
                    "createdAt": int(guild.created_at.timestamp() * 1000),
                })

            await with_client(run)

        case "get-channel":
            channel_id = opts.get("channel-id")
            if not channel_id:
                error("--channel-id is required")

            async def run(client: discord.Client):
                channel = await client.fetch_channel(int(channel_id))
                if not channel:
                    error("Channel not found")
                result = {
                    "id": str(channel.id),
                    "type": str(channel.type),
                }
                if hasattr(channel, "name"):
                    result["name"] = channel.name
                if hasattr(channel, "topic"):
                    result["topic"] = channel.topic
                if hasattr(channel, "category_id"):
                    result["parentId"] = str(channel.category_id) if channel.category_id else None
                output(result)

            await with_client(run)

        case _:
            error(
                f"Unknown command: {command}. "
                "Available: list-channels, read-messages, list-members, get-guild, get-channel"
            )


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except Exception as e:
        error(str(e))
